File: 11-02-2024_Cherry_pickup_II.cpp
class Solution {
public:

   int n, m;
   int t[71][71][71];

   int solve(vector<vector<int>> &grid , int r , int c1 , int c2){
       if(r>=m){
           return 0;
       }

       if(t[r][c1][c2]!=-1) 
       return t[r][c1][c2];

       int sum = grid[r][c1];
       if(c1!=c2) 
       sum+=grid[r][c2];

       int ans =0;

       for(int i=-1;i<=1;i++){
           for(int j=-1;j<=1;j++){
    
               int newc1=c1+i;
               int newc2=c2+j;

               if(newc2>=0 && newc2<n && newc1>=0 && newc1<n)
                ans =max(ans,solve(grid,r+1,newc1,newc2));
           }
       }

       return t[r][c1][c2]=sum+ans;
   }
   
    int cherryPickup(vector<vector<int>>& grid) {
         m = grid.size();
         n = grid[0].size();
        memset(t,-1,sizeof(t));

        return solve(grid,0,0,n-1);
    }
};

File: 11-02-2024_cherry_pickup_II.py
class Solution:
    def cherryPickup(self, grid: List[List[int]]) -> int:
        rows = len(grid)
        cols = len(grid[0])

        # Initialize the 3D DP array with -1
        DP = [[[-1] * cols for _ in range(cols)] for _ in range(rows)]

        # Base case: bottom row
        for j1 in range(cols):
            for j2 in range(cols):
                DP[rows - 1][j1][j2] = grid[rows - 1][j1]
                if j1 != j2:
                    DP[rows - 1][j1][j2] += grid[rows - 1][j2]

        # Iterate from the second-to-last row up to the top
        for i in range(rows - 2, -1, -1):
            for j1 in range(cols):
                for j2 in range(cols):
                    cherries = grid[i][j1]
                    if j1 != j2:
                        cherries += grid[i][j2]

                    # Iterate over the possible next positions for both robots
                    for nj1 in range(j1 - 1, j1 + 2):
                        for nj2 in range(j2 - 1, j2 + 2):
                            # Check if the next positions are within bounds
                            if 0 <= nj1 < cols and 0 <= nj2 < cols:
                                DP[i][j1][j2] = max(DP[i][j1][j2], cherries + DP[i + 1][nj1][nj2])

        # The answer is stored in DP[0][0][cols - 1]
        return DP[0][0][cols - 1]
